runtime/pkg/timeutil/timeutil.go

```diff
@@ -184,3 +184,69 @@ func AddTimeProto(to time.Time, tg runtimev1.TimeGrain, count int) time.Time {
 
 	return to
 }
+
+func OffsetTime(tm time.Time, tg TimeGrain, n int) time.Time {
+	switch tg {
+	case TimeGrainUnspecified:
+		return tm
+	case TimeGrainMillisecond:
+		return tm.Add(time.Duration(n) * time.Millisecond)
+	case TimeGrainSecond:
+		return tm.Add(time.Duration(n) * time.Second)
+	case TimeGrainMinute:
+		return tm.Add(time.Duration(n) * time.Minute)
+	case TimeGrainHour:
+		return tm.Add(time.Duration(n) * time.Hour)
+	case TimeGrainDay:
+		return tm.AddDate(0, 0, n)
+	case TimeGrainWeek:
+		return tm.AddDate(0, 0, n*7)
+	case TimeGrainMonth:
+		return tm.AddDate(0, n, 0)
+	case TimeGrainQuarter:
+		return tm.AddDate(0, n*3, 0)
+	case TimeGrainYear:
+		return tm.AddDate(n, 0, 0)
+	}
+
+	return tm
+}
+
+func CopyTimeTill(from, to time.Time, tg TimeGrain, tz *time.Location) time.Time {
```

Comment:
This function is slightly tricky to understand. Can you add a docstring? And maybe see if you can think of a more descriptive name? 

Not sure, but maybe `MergeComponents(target, source time.Time, until TimeGrain)`?

Reply:
Yeah that name is also good. Although `from` and `to` sound like a time range, which might be confusing (hence my suggestion of `source` and `target`, but open to suggestions).